src/app/cluster-building-blocks/QuieterReporting.h

```diff
@@ -0,0 +1,241 @@
+ * - When there is any increase or decrease in the estimated time remaining that was
+ *   due to progressing insight of the server's control logic
+ *   - Use SufficientChangePredicate version with an always-true predicate.
+ * - When it changes at a rate significantly different from one unit per second.
+ *   - Use SufficientChangePredicate version.
+ * Example usage in-situ:
+ *
+ * Class has:
+ *     QuieterReportingAttribute<uint8_t> mAttrib;
+ *
+ * Code at time of setting new value has:
+ *
+ *     uint8_t newValue = driver.GetNewValue();
+ *     auto now = SystemClock().GetMonotonicTimestamp();
+ *     if (mAttrib.SetValue(newValue, now) == AttributeDirtyState::kMustReport)
+ *     {
+ *         MatterReportingAttributeChangeCallback(path_for_attribute);
+ *     }
+ *
+ * @tparam T - the type of underlying numerical value that will be held by the class.
+ */
+template <typename T, std::enable_if_t<std::is_arithmetic<T>::value, bool> = true>
+class QuieterReportingAttribute
+{
+public:
+    explicit QuieterReportingAttribute(const Nullable<T> & initialValue) : mValue(initialValue), mLastDirtyValue(initialValue) {}
+
+    struct SufficientChangePredicateCandidate
+    {
+        // Timestamp of last time attribute was marked dirty.
+        Timestamp lastDirtyTimestamp;
+        // New (`now`) timestamp passed in `SetValue()`.
+        Timestamp nowTimestamp;
+        // Value last marked as dirty.
+        const Nullable<T> & lastDirtyValue;
+        // New value passed in `SetValue()`, to compare against lastDirtyValue for sufficient change if needed.
+        const Nullable<T> & newValue;
+    };
+
+    using SufficientChangePredicate = std::function<bool(const SufficientChangePredicateCandidate &)>;
+
+    /**
+     * @brief Factory to generate a functor for "attribute was last reported" at least `minimumDurationMillis` ago.
+     *
+     * @param minimumDurationMillis - number of millis needed since last marked as dirty before we mark dirty again.
+     * @return a functor usable for the `changedPredicate` arg of `SetValue()`
+     */
+    static SufficientChangePredicate
+    GetPredicateForSufficientTimeSinceLastDirty(System::Clock::Milliseconds64 minimumDurationMillis)
+    {
+        return [minimumDurationMillis](const SufficientChangePredicateCandidate & candidate) -> bool {
+            return (candidate.lastDirtyValue != candidate.newValue) &&
+                ((candidate.nowTimestamp - candidate.lastDirtyTimestamp) >= minimumDurationMillis);
+        };
+    }
+
+    /**
+     * @brief Factory to generate a functor that forces reportable now.
+     * @return a functor usable for the `changedPredicate` arg of `SetValue()`
+     */
+    static SufficientChangePredicate GetForceReportablePredicate()
+    {
+        return [](const SufficientChangePredicateCandidate & candidate) -> bool { return true; };
+    }
+
+    Nullable<T> value() const { return mValue; }
+    QuieterReportingPolicyFlags & policy() { return mPolicyFlags; }
```

Comment:
Do we really want policy to be randomly mutable?  I would expect it to be a constructor argument and immutable...

Reply:
I prefer being able to update the policy, and I don't think it's a big deal. Will leave like this.